### PR TITLE
Convert more Pow expressions to field elements

### LINKED ISSUE #13947: integrate(a**t /(a**t+1), t) sucess but a**(t-s)/(a**t+1) fail
No matter whether `conds='none'`.

``` python
import sympy as sm
print(sm.__version__)
# 1.1.1
from sympy import *
a, t, s = symbols('a t s')
f = a**t/(a**t+1)
integrate(f, t)
# >> log(a**t + 1)/log(a)
g = a**(t-s)/(a**t+1) 
integrate(g, t)
# hoping : a**(-s) log(a**t + 1)/log(a)
# but failed:
# CoercionFailed: can't convert exp(-s*log(a)) of type exp to ZZ(log(a),exp(s*log(a)))
```

And the error stack is
```
---------------------------------------------------------------------------
CoercionFailed                            Traceback (most recent call last)
<ipython-input-35-ff749dece9c9> in <module>()
----> 1 integrate(g, t)

/usr/lib/python3.5/site-packages/sympy/integrals/integrals.py in integrate(*args, **kwargs)
   1293     if isinstance(integral, Integral):
   1294         return integral.doit(deep=False, meijerg=meijerg, conds=conds,
-> 1295                              risch=risch, manual=manual)
   1296     else:
   1297         return integral

/usr/lib/python3.5/site-packages/sympy/integrals/integrals.py in doit(self, **hints)
    484                     function, xab[0],
    485                     meijerg=meijerg1, risch=risch, manual=manual,
--> 486                     conds=conds)
    487                 if antideriv is None and meijerg1 is True:
    488                     ret = try_meijerg(function, xab)

/usr/lib/python3.5/site-packages/sympy/integrals/integrals.py in _eval_integral(self, f, x, meijerg, risch, manual, conds)
    778         if risch is not False:
    779             try:
--> 780                 result, i = risch_integrate(f, x, separate_integral=True, conds=conds)
    781             except NotImplementedError:
    782                 pass

/usr/lib/python3.5/site-packages/sympy/integrals/risch.py in risch_integrate(f, x, extension, handle_first, separate_integral, rewrite_complex, conds)
   1727         fa, fd = fa.cancel(fd, include=True)
   1728         if case == 'exp':
-> 1729             ans, i, b = integrate_hyperexponential(fa, fd, DE, conds=conds)
   1730         elif case == 'primitive':
   1731             ans, i, b = integrate_primitive(fa, fd, DE)

/usr/lib/python3.5/site-packages/sympy/integrals/risch.py in integrate_hyperexponential(a, d, DE, z, conds)
   1457 
   1458     g1, h, r = hermite_reduce(a, d, DE)
-> 1459     g2, b = residue_reduce(h[0], h[1], DE, z=z)
   1460     if not b:
   1461         i = cancel(a.as_expr()/d.as_expr() - (g1[1]*derivation(g1[0], DE) -

/usr/lib/python3.5/site-packages/sympy/integrals/risch.py in residue_reduce(a, d, DE, z, invert)
   1215     z = z or Dummy('z')
   1216     a, d = a.cancel(d, include=True)
-> 1217     a, d = a.to_field().mul_ground(1/d.LC()), d.to_field().mul_ground(1/d.LC())
   1218     kkinv = [1/x for x in DE.T[:DE.level]] + DE.T[:DE.level]
   1219 

/usr/lib/python3.5/site-packages/sympy/polys/polytools.py in mul_ground(f, coeff)
   1204         """
   1205         if hasattr(f.rep, 'mul_ground'):
-> 1206             result = f.rep.mul_ground(coeff)
   1207         else:  # pragma: no cover
   1208             raise OperationNotSupported(f, 'mul_ground')

/usr/lib/python3.5/site-packages/sympy/polys/polyclasses.py in mul_ground(f, c)
    405     def mul_ground(f, c):
    406         """Multiply ``f`` by a an element of the ground domain. """
--> 407         return f.per(dmp_mul_ground(f.rep, f.dom.convert(c), f.lev, f.dom))
    408 
    409     def quo_ground(f, c):

/usr/lib/python3.5/site-packages/sympy/polys/domains/domain.py in convert(self, element, base)
    156                     pass
    157 
--> 158         raise CoercionFailed("can't convert %s of type %s to %s" % (element, type(element), self))
    159 
    160     def of_type(self, element):

CoercionFailed: can't convert exp(-s*log(a)) of type exp to ZZ(log(a),exp(s*log(a)))
```

### PR DESCRIPTION
**Pull request by normalhuman**
#### References to other Issues or PRs

Fixes #13947

#### Brief description of what is fixed or changed

The `FracField._rebuild_expr` method did not try hard enough to express a given Pow in terms of field generators: it just expected the exponent to be integer, and the base be in the field. This missed, for example, that `Z(2**pi)` contains `2**(-pi)`. Now the Pow and ExpBase instances are treated more carefully, by comparing their bases to the bases of Pow or ExpBase among generators, and seeking an integer ratio of exponents.

### CODE REVIEW

**Comment by jksuom** [comment]
```
Python 2.7.6 (default, Nov 23 2017, 15:49:48) 
[GCC 4.8.4] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> from sympy.polys.fields import *
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "sympy/__init__.py", line 60, in <module>
    from .polys import *
  File "sympy/polys/__init__.py", line 69, in <module>
    from . import fields
  File "sympy/polys/fields.py", line 209
    powers = tuple((gen, *gen.as_base_exp()) for gen in mapping.keys()
                         ^
SyntaxError: invalid syntax
```
This does not seem to work in Python 2.

**Comment by ylemkimon** [comment]
Using the unpacking operator(`*`) in other than function arguments is only possible in Python 3.5+.

File: sympy/polys/fields.py [hunk 1]
```diff
@@ -214,16 +218,24 @@ def _rebuild(expr):
...
-                    return domain.convert(expr)
-                except CoercionFailed:
-                    if not domain.is_Field and domain.has_assoc_Field:
-                        return domain.get_field().convert(expr)
-                    else:
-                        raise
+            elif expr.is_Pow or isinstance(expr, ExpBase):
+                b, e = expr.as_base_exp()
+                # look for bg**eg whose integer power may be b**e
+                choices = tuple((gen, bg, eg) for gen, bg, eg in powers
```
Comment (jksuom) [inline sympy/polys/fields.py]: This could have `for gen, (bg, eg) in powers` if `*` is omitted from `*gen.as_base_exp()` above.

**Comment by jksuom** [comment]
Looks like connection problem. Restarted.

File: sympy/polys/fields.py [hunk 2]
```diff
@@ -214,16 +218,24 @@ def _rebuild(expr):
...
-                    if not domain.is_Field and domain.has_assoc_Field:
-                        return domain.get_field().convert(expr)
-                    else:
-                        raise
+            elif expr.is_Pow or isinstance(expr, ExpBase):
+                b, e = expr.as_base_exp()
+                # look for bg**eg whose integer power may be b**e
+                choices = tuple((gen, bg, eg) for gen, (bg, eg) in powers
+                    if bg == b and Mod(e, eg) == 0)
+                if len(choices) > 0:
```
Comment (jksuom) [inline sympy/polys/fields.py]: This could be `if choices:`.

**Comment by jksuom** [comment]
Looks good.

**Pull request merged**